### ISSUE TITLE
SROAP02016 ClassNotFoundException for public enums

### ISSUE THREAD

**Issue by scutaru** [smallrye/smallrye-open-api#1508]
When I deploy my application in Wildfly 28.0.1 I get the following type of warnings for all the enums:

_2023-07-04 13:37:38,516 WARN  [MSC service thread 1-8] [exchangeID=, sessionID=] (io.smallrye.openapi.runtime.io) SROAP02016: Failed to read enumeration values from enum com.example.ErrorSourceEnum method toString with `@JsonValue`: java.lang.ClassNotFoundException: com.example.ErrorSourceEnum from [Module "io.smallrye.openapi" version 3.3.2 from local module loader @21ba0741 (finder: local module finder @58ce9668 (roots: /opt/jboss/wildfly-28.0.1.Final/modules,/opt/jboss/wildfly-28.0.1.Final/modules/system/layers/base))]_

The enum looks like below:
```
package com.example;

import com.fasterxml.jackson.annotation.JsonValue;

public enum ErrorSourceEnum {
    ISSUER1("ISSUER 1"),

    ISSUER2("ISSUER 2");

    private String value;

    ErrorSourceEnum(String value) {
        this.value = value;
    }

    @Override
    @JsonValue
    public String toString() {
        return value;
    }
}
```

Any idea what could be wrong ?

**Comment by MikeEdgar** [comment]
Hi @scutaru , the scanner is attempting to get the value using the method annotated with `@JsonValue`, but the enum isn't available to the scanner module. Are you using a `module-info` in your application?

**Comment by scutaru** [comment]
No, I'm not

**Comment by scutaru** [comment]
I'm migrating it from Wildfly 26 to Wildfly 28. But I just checked and the same warnings were displayed in Wildfly 26 too:

_2023-07-05 14:29:34,638 WARN  [MSC service thread 1-7] [exchangeID=, sessionID=] (io.smallrye.openapi.runtime.io) SROAP02016: Failed to read enumeration values from enum com.example.ErrorSourceEnum method toString with `@JsonValue`: java.lang.ClassNotFoundException: com.example.ErrorSourceEnum from [Module "io.smallrye.openapi" from local module loader @5a9f4771 (finder: local module finder @282cb7c7 (roots: /opt/jboss/wildfly-26.1.3.Final/modules,/opt/jboss/wildfly-26.1.3.Final/modules/system/layers/base))]_

So it's not that urgent anymore, but I'd still like to know how to fix it.

**Comment by MikeEdgar** [comment]
@pferraro , do you have any insight on what could be the issue? I assume Wildfly modules cannot reflect on the application classes? Is there any work-around?

**Comment by pferraro** [comment]
@scutaru Do you have a full stack trace?  Smallrye is trying to load the enum class from its own module rather than the deployment module, which is certainly wrong.  @MikeEdgar Do you know offhand where this might be happening?

**Comment by MikeEdgar** [comment]
@pferraro , this is the site where the class is loaded: https://github.com/smallrye/smallrye-open-api/blob/c728c55885be0371bf9f79c6c004fbc8e9145815/core/src/main/java/io/smallrye/openapi/runtime/io/schema/SchemaFactory.java#L570

The class loader that is used would have been set by the client, in this case the integration code in Wildfly. If Wildfly is using one of the `OpenApiProcessor#bootstrap` overloads, it may be one that doesn't pass a `ClassLoader`.

**Comment by pferraro** [comment]
@MikeEdgar Thanks.  The problem, as I see it, is that the ClassLoader provided to OpenApiAnnotationScanner serves two very different roles: loading AnnotationScanner instances (via AnnotationScannerFactory) and loading application classes.
I can certainly use an aggregating ClassLoader instead - but it would be much cleaner if these could be specified independently, or, even better, allow the platform to bypass the service loading required by AnnotationScannerFactory.

**Comment by MikeEdgar** [comment]
> it would be much cleaner if these could be specified independently, or, even better, allow the platform to bypass the service loading required by AnnotationScannerFactory

Yes, I agree that something needs to change with this. For the second option you gave to bypass the service loading, are you suggesting that the platform would provide pre-constructed `AnnotationScanner` instances?

**Comment by pferraro** [comment]
> > it would be much cleaner if these could be specified independently, or, even better, allow the platform to bypass the service loading required by AnnotationScannerFactory
> 
> Yes, I agree that something needs to change with this. For the second option you gave to bypass the service loading, are you suggesting that the platform would provide pre-constructed `AnnotationScanner` instances?

Either that, or allow it to provide its own AnnotationScannerFactory implementation (assuming this was an interface, or even just a Supplier<Iterable<AnnotationScanner>>)

**Comment by pferraro** [comment]
I opened https://issues.redhat.com/browse/WFLY-18224 and submitted https://github.com/wildfly/wildfly/pull/17004/commits/25dd6bd7c0d166a05fea3546ecb0710fc05434ec to fix the issue until given the limitations of the current SPIs.

**Comment by scutaru** [comment]
> @scutaru Do you have a full stack trace? Smallrye is trying to load the enum class from its own module rather than the deployment module, which is certainly wrong. @MikeEdgar Do you know offhand where this might be happening?

Unfortunately that's the only problem that I see in the logs, nothing else.
